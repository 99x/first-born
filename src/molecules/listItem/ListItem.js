import React, { Component } from "react";
import {
    View,
    StyleSheet,
    TouchableOpacity,
    Image,
    Platform
} from "react-native";
import PropTypes from "prop-types";

import { Text } from "../../atoms/Text";
import { Icon } from "../../atoms/Icon";
import { commonColors } from "../../utils/color";
import { deviceVariables } from "../../variables/deviceVariables";

export class ListItem extends Component {
    constructor(props) {
        super(props);
        this.state = {
            expanded: false
        };
    }

    render() {
        const {
            title,
            description,
            image,
            onPress,
            block,
            backgroundColor,
            secondary,
            ...otherProps
        } = this.props;
        const { expanded } = this.state;

        const children = React.Children.map(this.props.children, child =>
            child
                ? React.cloneElement(child, { ...child.props, secondary: true })
                : null
        );

        const listItemBackColor = secondary
            ? commonColors.secondaryBackground
            : backgroundColor;

        const listItemStyle = block
            ? [styles.containerBlock, { backgroundColor: listItemBackColor }]
            : [styles.container, { backgroundColor: listItemBackColor }];

        const onListItemTap = children
            ? this.expandListItem
            : onPress
            ? onPress
            : undefined;

        return (
            <View style={{ width: "100%" }}>
                <TouchableOpacity
                    style={listItemStyle}
                    disabled={!onListItemTap}
                    onPress={onListItemTap}
                    {...otherProps}
                >
                    {image && (
                        <View style={styles.imageContainer}>
                            <Image style={styles.image} {...image} />
                        </View>
                    )}
                    <View style={styles.textContainer}>
                        <Text size="h6" color={commonColors.darkGrey} bold>
                            {title}
                        </Text>
                        {description && (
                            <Text
                                size="sub_heading"
                                color={commonColors.lightGrey}
                            >
                                {description}
                            </Text>
                        )}
                    </View>
                    <View style={styles.imageContainer}>
                        {children && (
                            <Icon
                                name={
                                    expanded
                                        ? Platform.OS === "ios"
                                            ? "arrow-up"
                                            : "arrow-dropup"
                                        : Platform.OS === "ios"
                                        ? "arrow-down"
                                        : "arrow-dropdown"
                                }
                                size={20}
                                color={commonColors.inputGrey}
                            />
                        )}
                    </View>
                </TouchableOpacity>
                {expanded && children}
            </View>
        );
    }

    expandListItem = () => {
        this.setState({ expanded: !this.state.expanded });
    };
}

ListItem.propTypes = {
    title: PropTypes.string.isRequired,
    description: PropTypes.string,
    image: PropTypes.shape({
        ...Image.propTypes
    }),
    block: PropTypes.bool,
    backgroundColor: PropTypes.string,
    secondary: PropTypes.bool,
    ...TouchableOpacity.propTypes
};

ListItem.defaultProps = {
    backgroundColor: commonColors.white,
    secondary: false,
    block: false
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: "row",
        ...Platform.select({
            android: {
                borderWidth: 0.5,
                borderColor: commonColors.lightGrey
            },
            ios: {
                borderTopWidth: 0.5,
                borderBottomWidth: 0.5,
                borderTopColor: commonColors.lightGrey,
                borderBottomColor: commonColors.lightGrey
            }
        }),

        paddingHorizontal: 5
    },
    containerBlock: {
        width: deviceVariables.width,
        flex: 1,
        flexDirection: "row",
        borderTopWidth: 0.5,
        borderBottomWidth: 0.5,
        borderTopColor: commonColors.lightGrey,
        borderBottomColor: commonColors.lightGrey,
        paddingHorizontal: 15
    },
    textContainer: {
        flex: 3,
        flexDirection: "column",
        padding: 15,
        alignItems: "flex-start"
    },
    imageContainer: {
        flex: 1,
        justifyContent: "center",
        alignItems: "center"
    },
    image: {
        height: 50,
        width: 50,
        borderRadius: 50 / 2
    }
});
