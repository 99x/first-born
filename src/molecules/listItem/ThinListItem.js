import React, { Component } from "react";
import {
    View,
    StyleSheet,
    TouchableOpacity,
    Image,
    Platform
} from "react-native";
import PropTypes from "prop-types";

import { Text } from "../../atoms/Text";
import { Icon } from "../../atoms/Icon";
import { commonColors } from "../../utils/color";
import { deviceVariables } from "../../variables/deviceVariables";

export class ThinListItem extends Component {
    render() {
        const {
            title,
            description,
            image,
            icon,
            backgroundColor,
            arrow,
            ...otherProps
        } = this.props;

        const listItemStyle = [styles.container, { backgroundColor }];

        return (
            <TouchableOpacity style={listItemStyle} {...otherProps}>
                {image && (
                    <View style={styles.imageContainer}>
                        <Image style={styles.image} {...image} />
                    </View>
                )}
                {icon && (
                    <View style={styles.imageContainer}>
                        <Icon size={20} {...icon} />
                    </View>
                )}
                <View style={styles.textContainer}>
                    <Text color={commonColors.darkGrey} bold>
                        {title}
                    </Text>
                    {description && (
                        <Text size="footnote" color={commonColors.lightGrey}>
                            {description}
                        </Text>
                    )}
                </View>
                {arrow && (
                    <View style={styles.imageContainer}>
                        <Icon
                            type={
                                Platform.OS === "ios" ? "ionicon" : "material"
                            }
                            name={
                                Platform.OS === "ios"
                                    ? "arrow-forward"
                                    : "keyboard-arrow-right"
                            }
                            size={20}
                            color={commonColors.inputGrey}
                        />
                    </View>
                )}
            </TouchableOpacity>
        );
    }
}

ThinListItem.propTypes = {
    title: PropTypes.string.isRequired,
    description: PropTypes.string,
    image: PropTypes.shape({
        ...Image.propTypes
    }),
    icon: PropTypes.shape({
        ...Icon.propTypes
    }),
    backgroundColor: PropTypes.string,
    arrow: PropTypes.bool,
    ...TouchableOpacity.propTypes
};

ThinListItem.defaultProps = {
    backgroundColor: commonColors.white,
    arrow: false
};

const styles = StyleSheet.create({
    container: {
        width: deviceVariables.width,
        flex: 1,
        flexDirection: "row",
        borderTopWidth: 0.5,
        borderBottomWidth: 0.5,
        borderTopColor: commonColors.lightGrey,
        borderBottomColor: commonColors.lightGrey,
        paddingLeft: 5,
        paddingRight: 10
    },
    textContainer: {
        flex: 4,
        flexDirection: "column",
        paddingVertical: 10,
        paddingHorizontal: 15,
        alignItems: "flex-start"
    },
    imageContainer: {
        justifyContent: "center",
        alignItems: "center"
    },
    image: {
        height: 30,
        width: 30,
        borderRadius: 30 / 2
    }
});
