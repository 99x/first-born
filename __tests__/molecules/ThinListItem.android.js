import "react-native";
import React from "react";
import renderer from "react-test-renderer";
import { ThinListItem } from "../../src/molecules/listItem/ThinListItem";

jest.mock("Platform", () => {
    const Platform = require.requireActual("Platform");
    Platform.OS = "android";
    return Platform;
});

it("renders thin list item with heading", () => {
    const tree = renderer.create(<ThinListItem title="Heading" />).toJSON();
    expect(tree).toMatchSnapshot();
});

it("renders thin list item with heading and title", () => {
    const tree = renderer
        .create(<ThinListItem title="Heading" description="description" />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it("renders thin list item with heading, title and image", () => {
    const tree = renderer
        .create(
            <ThinListItem
                title="Heading"
                description="description"
                image={{ source: require("../../firstBorn-logo.png") }}
            />
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it("renders thin list item with heading, title and arrow", () => {
    const tree = renderer
        .create(
            <ThinListItem title="Heading" description="description" arrow />
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it("renders thin list item with heading, title, image and arrow", () => {
    const tree = renderer
        .create(
            <ThinListItem
                title="Heading"
                description="description"
                image={{ source: require("../../firstBorn-logo.png") }}
                arrow
            />
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it("renders colored thin list item with heading and title", () => {
    const tree = renderer
        .create(
            <ThinListItem
                title="Heading"
                description="description"
                color="aliceblue "
            />
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
