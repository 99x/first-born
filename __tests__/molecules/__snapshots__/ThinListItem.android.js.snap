// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders colored thin list item with heading and title 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#FFF",
      "borderBottomColor": "rgba(33, 33, 33, 0.4)",
      "borderBottomWidth": 0.5,
      "borderTopColor": "rgba(33, 33, 33, 0.4)",
      "borderTopWidth": 0.5,
      "flex": 1,
      "flexDirection": "row",
      "opacity": 1,
      "paddingLeft": 5,
      "paddingRight": 10,
      "width": 750,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flex": 4,
        "flexDirection": "column",
        "paddingHorizontal": 15,
        "paddingVertical": 10,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.87)",
            "fontSize": 16,
            "fontWeight": "bold",
            "textAlign": "left",
          },
        ]
      }
    >
      Heading
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.4)",
            "fontSize": 13,
            "fontWeight": "normal",
            "textAlign": "left",
          },
        ]
      }
    >
      description
    </Text>
  </View>
</View>
`;

exports[`renders thin list item with heading 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#FFF",
      "borderBottomColor": "rgba(33, 33, 33, 0.4)",
      "borderBottomWidth": 0.5,
      "borderTopColor": "rgba(33, 33, 33, 0.4)",
      "borderTopWidth": 0.5,
      "flex": 1,
      "flexDirection": "row",
      "opacity": 1,
      "paddingLeft": 5,
      "paddingRight": 10,
      "width": 750,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flex": 4,
        "flexDirection": "column",
        "paddingHorizontal": 15,
        "paddingVertical": 10,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.87)",
            "fontSize": 16,
            "fontWeight": "bold",
            "textAlign": "left",
          },
        ]
      }
    >
      Heading
    </Text>
  </View>
</View>
`;

exports[`renders thin list item with heading and title 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#FFF",
      "borderBottomColor": "rgba(33, 33, 33, 0.4)",
      "borderBottomWidth": 0.5,
      "borderTopColor": "rgba(33, 33, 33, 0.4)",
      "borderTopWidth": 0.5,
      "flex": 1,
      "flexDirection": "row",
      "opacity": 1,
      "paddingLeft": 5,
      "paddingRight": 10,
      "width": 750,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flex": 4,
        "flexDirection": "column",
        "paddingHorizontal": 15,
        "paddingVertical": 10,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.87)",
            "fontSize": 16,
            "fontWeight": "bold",
            "textAlign": "left",
          },
        ]
      }
    >
      Heading
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.4)",
            "fontSize": 13,
            "fontWeight": "normal",
            "textAlign": "left",
          },
        ]
      }
    >
      description
    </Text>
  </View>
</View>
`;

exports[`renders thin list item with heading, title and arrow 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#FFF",
      "borderBottomColor": "rgba(33, 33, 33, 0.4)",
      "borderBottomWidth": 0.5,
      "borderTopColor": "rgba(33, 33, 33, 0.4)",
      "borderTopWidth": 0.5,
      "flex": 1,
      "flexDirection": "row",
      "opacity": 1,
      "paddingLeft": 5,
      "paddingRight": 10,
      "width": 750,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flex": 4,
        "flexDirection": "column",
        "paddingHorizontal": 15,
        "paddingVertical": 10,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.87)",
            "fontSize": 16,
            "fontWeight": "bold",
            "textAlign": "left",
          },
        ]
      }
    >
      Heading
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.4)",
            "fontSize": 13,
            "fontWeight": "normal",
            "textAlign": "left",
          },
        ]
      }
    >
      description
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "rgba(33, 33, 33, 0.5)",
            "fontSize": 20,
          },
          undefined,
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      îŒ•
    </Text>
  </View>
</View>
`;
